### PR TITLE
fix #5214: explicitly omitting nulls by default

### PR DESCRIPTION
**Pull request by shawkins**
## Description

Resolves the discussion on #5214 - we'll treat the 6.6 / 6.7 behavior as a regression and switch to explicitly omitting nulls by default.

## Type of change

 - [x] Bug fix (non-breaking change which fixes an issue)
 - [ ] Feature (non-breaking change which adds functionality)
 - [ ] Breaking change (fix or feature that would cause existing functionality to change
 - [ ] Chore (non-breaking change which doesn't affect codebase;
   test, version modification, documentation, etc.)

## Checklist
 - [x] Code contributed by me aligns with current project license: [Apache 2.0](https://www.apache.org/licenses/LICENSE-2.0)
 - [x] I Added [CHANGELOG](https://github.com/fabric8io/kubernetes-client/blob/master/CHANGELOG.md) entry regarding this change
 - [x] I have implemented unit tests to cover my changes
 - [ ] I have added/updated the [javadocs](https://www.javadoc.io/doc/io.fabric8/kubernetes-client/latest/index.html) and other [documentation](https://github.com/fabric8io/kubernetes-client/blob/master/doc/CHEATSHEET.md) accordingly
 - [ ] No new bugs, code smells, etc. in [SonarCloud](https://sonarcloud.io/dashboard?id=fabric8io_kubernetes-client) report
 - [ ] I tested my code in Kubernetes
 - [ ] I tested my code in OpenShift

### CODE REVIEW

**Comment by manusa** [comment]
Unrelated to this change (I think), but worth noting and related to the general functionality, especially for patch scenarios.

Our Kubernetes Resource objects are currently initialized with empty arrays (I think we did add this behavior a while ago). This poses a problem when using builder to generate objects that will be sent to the API server in a JSON_MERGE (`application/merge-patch+json`) PATCH request. If the remote object has an entry in the array, since the sent patch JSON contains the field with an empty array, the remote field will be overwritten and cleared.

**Comment by shawkins** [comment]
> Unrelated to this change (I think), but worth noting and related to the general functionality, especially for patch scenarios.
> 
> Our Kubernetes Resource objects are currently initialized with empty arrays (I think we did add this behavior a while ago). This poses a problem when using builder to generate objects that will be sent to the API server in a JSON_MERGE (`application/merge-patch+json`) PATCH request. If the remote object has an entry in the array, since the sent patch JSON contains the field with an empty array, the remote field will be overwritten and cleared.

@manusa for the built-in resources where we initialize empty collections there is also the annotation: @JsonInclude(JsonInclude.Include.NON_EMPTY) - so the expectation moving forward is that neither a null nor an empty collection will be included in the serialization by default.  So I don't think the situation you are describing currently occurs with the built-in resources.  However the opposite problem exists - this was discussed on #4606 but wasn't addressed.

File: CHANGELOG.md [hunk 1]
```diff
@@ -3,6 +3,7 @@
 ### 6.8-SNAPSHOT
```
Comment (andreaTP) [inline CHANGELOG.md:3]: Given the discussion, this fix should probably be included in a 6.7 patch release.

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesSerialization.java [hunk 2]
```diff
@@ -89,6 +91,8 @@ public KubernetesSerialization(ObjectMapper mapper, boolean searchClassloaders)
   protected void configureMapper(ObjectMapper mapper) {
     mapper.registerModules(new JavaTimeModule(), unmatchedFieldTypeModule);
     mapper.disable(DeserializationFeature.FAIL_ON_INVALID_SUBTYPE);
+    // omit null fields, but keep null map values
+    mapper.setDefaultPropertyInclusion(JsonInclude.Value.construct(Include.NON_NULL, Include.ALWAYS));
```
Comment (andreaTP) [inline kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesSerialization.java:95]: Note for the future, is it possible to let the user customize those options completely?

**Review by andreaTP — APPROVED** [review]
Super late at this party, but thanks a lot for the changes @shawkins !

I agree this is the right direction, a couple of minor notes.

**Comment by manusa** [comment]
> So I don't think the situation you are describing currently occurs with the built-in resources

It does occur (for sure at least on 6.6, but I guess this is the same on master and earlier versions too) for `DeploymentConfig`, for example.

I think this has to do with the `javaOmitEmpty` JSON Schema configuration that's computed using the `getPropertyDescriptor` function:
 
https://github.com/fabric8io/kubernetes-client/blob/60230428cb90b4e632913e85925536f9537b1280/kubernetes-model-generator/pkg/schemagen/generate.go#L313

And the fact that builders are initialized with empty collections.

The following link contains a concrete example of why this is a problem:
https://github.com/manusa/yakd/blob/6d942632ee3ad5e738ba31c0b2fe45bac99f3c99/src/main/java/com/marcnuri/yakd/quickstarts/dashboard/deploymentconfigs/DeploymentConfigService.java#L90-L97

Basically when sending the patch, the JSON sent to the server contains something like:
```
{"spec": {"replicas": 1, "triggers": []}} 
```

Since this is a JSON_MERGE patch, the `triggers` are wiped out from the server resource.

**Comment by shawkins** [comment]
> It does occur (for sure at least on 6.6, but I guess this is the same on master and earlier versions too) for DeploymentConfig, for example.

So yes that would be a bug with the omitEmpty logic.  I recall specifically adding a workaround for all maps in Fabric8DefaultRule but wasn't aware we had an inconsistency with some lists (others lists of either simple or complex types like objectmeta.finalizers or ownerreferences do end up annotated appropriately).  I'd be fine with another override in Fabric8DefaultRule to act as a catch all.

> And the fact that builders are initialized with empty collections.

Builders make quite a few assumptions about null / empty - the initialization, no null map values, and no null complex type collection values, and probably more...  At least the builder behavior to default to an empty collection was part of the motivation of the change to initialize all of pojos with empty collections.

File: kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesSerialization.java (hunk 2 above)
Comment (shawkins) [reply kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesSerialization.java:95]: Only in an advanced scenario usage scenario.  The configureMapper method is protected and may be overriden if someone wants a create a custom KubernetesSerialization subclass.  However I wanted to work through the quarkus integration and especially a plan to minimize the number of classes referenced by the kubernetesdeserializer - I'm thinking we need provide a way to lazily load - before open things up even more.

Comment (andreaTP) [reply kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/KubernetesSerialization.java:95]: Great idea!

**Comment by shawkins** [comment]
@manusa the rationale for the inconsistency in the deploymentconfig spec triggers is that they are not marked as omitempty upstream: https://github.com/openshift/api/blob/master/apps/v1/types.go#L63

As you point out our builders are opionated about treating an empty and null collection the same - and we're already initializing the trigger list to empty, so I would be fine to just force the annotation.  That can either be done in the generator or via the default rule.  Do you want that as part of this pr?
